app/client/ui/Experiments.ts

```diff
@@ -0,0 +1,189 @@
+
+  constructor(topAppModel: TopAppModel) {
+    super();
+    this._topAppModel = topAppModel;
+  }
+
+  public isEnabled(experiment: Experiment) {
+    const experimentState = this._getExperimentState(experiment);
+    return experimentState.enabled;
+  }
+
+  /**
+   * Returns whether or not the user wants to show the experiments modal.
+   */
+  public isRequested() {
+    const urlExperiment = this.getCurrentRequest();
+    return urlExperiment && this._isSupported(urlExperiment);
+  }
+
+  /**
+   * Returns the experiment that the user wants to show the modal for.
+   */
+  public getCurrentRequest() {
+    const searchParams = new URLSearchParams(G.window.location.search);
+    return searchParams.get(EXPERIMENT_URL_PARAM);
+  }
+
+  /**
+   * Shows the modal for the given experiment, allowing the user to enable or disable it.
+   */
+  public showModal(experiment: string) {
+    if (!this._isSupported(experiment)) {
+      return;
+    }
+
+    // if the app is not initialized, wait and retry, we need the current user info
+    const appObs = this._topAppModel.appObs;
+    if (appObs.get() === null) {
+      appObs.addListener((app, prevApp) => {
+        if (app && prevApp === null) {
+          this.showModal(experiment);
+        }
+      });
+      return;
+    }
+
+    const experimentState = this._getExperimentState(experiment);
+
+    const alreadyEnabled = experimentState.enabled;
+
+    const experimentLabel = EXPERIMENTS[experiment as keyof typeof EXPERIMENTS]();
+    const hasConfirmedModal = Observable.create(this, false);
+    const experimentUrl = new URL(getGristConfig().homeUrl || window.location.href);
+    experimentUrl.searchParams.set(EXPERIMENT_URL_PARAM, experiment);
+    const urlBlock = `
+      <a class="${cssLink.className}" href="${experimentUrl}">
+        ${experimentUrl}
+      </a>
+    `;
+
+    if (!hasConfirmedModal.get()) {
+      confirmModal(
+        t('Experimental feature'),
+        alreadyEnabled ? t('Disable feature') : t('Enable feature'),
+        () => {
+          this._setExperimentState(experiment, !alreadyEnabled);
+          hasConfirmedModal.set(true);
+        },
+        {
+          explanation: cssWrapper(
+            dom('p', dom.cls(cssWrapper.className), (el) => {
+              el.innerHTML = t(
+                alreadyEnabled
+                  ? 'You are about to disable this experimental feature: {{- experiment}}'
+                  : 'You are about to enable this experimental feature: {{- experiment}}',
+                {experiment: `
+                  <strong>
+                    ${experimentLabel}
+                  </strong>`}
+              );
+            }),
+            !alreadyEnabled ? dom('p', t("Don't worry, you can disable it later if needed.")) : null
+          )
+        }
+      );
+    }
+
+    // If the user just confirmed the feature toggle, feedback the change and force a page reload
+    hasConfirmedModal.addListener((val) => {
```

Comment:
I generally recommend avoiding `addListener`, and in this case I don't think you even need an observable. You could have a method `_showConfirmation()` that shows the first dialog, and `_showReloadPrompt()` for the second dialog, and instead of calling `hasConfirmedModal.set(true)`, you could just call `this._showReloadPrompt()`. (I may be missing some subtlety, but it looks like that's all that's happening.)

Reply:
you are totally right, don't know what I was thinking here :)